### DISCUSSION TITLE
Question: Difference between CollectAndExecuteWorkflowsAsync and CollectAndDispatchWorkflowsAsync

### DISCUSSION THREAD

**Discussion by alkampfergit** [elsa-workflows/elsa-core#3864] [Q&A] [answered]
Hi,

I have some components that starts workflows upon various custom trigger, and I wonder what is the exact implication of using CollectAndExecuteWorkflowsAsync  vs CollectAndDispatchWorkflowsAsync.

From what I understand CollectAndExecuteWorkflowsAsync will block the thread until the workflow is completed (or suspended?) while CollectAndDispatchWorkflowsAsync will simply starts the workflow and not blocking the calling thread waiting for the workflow to finish. 

Is this the only difference or there are some other caveats I need to be consider?

Thanks for any help.

**Comment by sfmskywalker** [comment, marked as answer]
Hi @alkampfergit 

You're spot on; the only difference is from where the workflow is executed. To recap:

- Execute: this is a blocking call (if you `await` the task) the workflow is executed until its end or until a blocking activity is encountered.
- Dispatch: this is a non-blocking call, because it schedules a message onto the queue and then returns. The workflow will be executed by the queue worker, which happens asynchronously, potentially in another container instance in a multi-node cluster.

There are some caveats to this. For example, if your workflow contains activities that rely on HTTP context, such as the HTTP Response activity, those will no longer work, since the queue worker executes somewhere in a background thread with no HTTP context.

Another thing to be aware of is that, with Elsa 2, sometimes you want to use Dispatch and not Execute when using signals.
For example, if you have Workflow A that sends a signal to Workflow B, which then sends a signal back to Workflow A, you are deadlocking both workflows, because while Workflow A is executing the signal, it is inside a lock. When Workflow B tries to send the signal, Workflow A can't handle it because it is in a lock. Using a Dispatch approach to send a signal, on the other hand doesn't have this issue.
